micronota/bfillings/prodigal.py

```diff
@@ -6,251 +19,196 @@
+                help='Select procedure (single or meta).  Default is single.'),
+    OptionParam('-c', help='Closed ends.  Do not allow genes to run off edges.'),
+    OptionParam('-v', help='Print version number and exit.'),
+    OptionParam('-q', help='Run quietly (suppress normal stderr output).'),
+    OptionParam('-h', help='Print help menu and exit.'),
+    OptionParam('-n', help='Bypass Shine-Dalgarno trainer and force a full motif scan.')]
+
+
+def run(query, out_dir, **kwargs):
+    '''Run prodigal for gene prediction.
+
+    Notes
+    -----
+    It will create 3 output files with prefix of "prodigal" in "out_dir" folder:
+      1. the annotation file (GFF3 file by default)
+      2. the nucleotide sequences for each predicted gene
+         with file suffix of .fna.
+      3. the protein sequence translated from each gene
+         with file suffix of .faa.
+
+    Parameters
+    ----------
+    query : str
+        input file path of sequence
+    out_dir : str
+        output dir
+    kwargs : dict
+        keyword arguments for Prodigal.
+
+    Returns
+    -------
+    `Dumpling`
+    '''
+    logger = getLogger(__name__)
+
+    makedirs(out_dir, exist_ok=True)
+
+    prodigal = Dumpling('prodigal', params=Parameters(*params),
+                        version='v2.6.3', url='https://github.com/hyattpd/Prodigal')
+    # set default output to gff and run mode to draft genome
+    prodigal.update(query=query, fmt='gff', mode='single')
+    # update with kwargs
+    prodigal.update(**kwargs)
+
+    suffices = {
+        '-a': 'faa',
+        # output file of nucleotide sequences of genes
+        '-d': 'fna',
+        '-o': prodigal.params['fmt'].value}
+
+    for flag in ['-a', '-d', '-o']:
+        p = prodigal.params[flag]
+        if p.is_off():
+            p.on(join(out_dir, 'prodigal.{}'.format(suffices[flag])))
+
+    logger.info('Running CDS prediction {}'.format(prodigal.command))
+    prodigal(stdout=join(out_dir, 'prodigal.log'))
+    return prodigal
+
+
+def parse_result(self, res, which='-a'):
+    '''Parse gene prediction result from ``Prodigal``.
+
+    It is parsed into a dict consumable by
+    ``skbio.metadata.IntervalMetadata``.
+
+    Parameters
+    ----------
+    res : burrito.util.CommandLineAppResult
+    which : which output to parse
+    Returns
+    -------
+    ``skbio.metadata.IntervalMetadata``
+    '''
+    # make sure to move to the beginning of the file.
+    if which == '-a':
+        return _parse_faa(res[which])
+    elif which == '-o':
+        return _parse_records(res[which], _parse_single_record)
+
+
+def _parse_single_record(chunks):
+    '''Parse single record of Prodigal GenBank output.
+
+    Parameters
+    ----------
+    chunks : list of str
+        a list of lines of the record to parse.
+
+    Yields
+    ------
+    dict passable to ``skbio.metadata.IntervalMetadata``
+    '''
+    # get the head line
+    head = chunks[0]
+    _, description = head.split(None, 1)
```

Comment:
That's cool!

On Thu, Apr 28, 2016 at 2:44 PM, Zech Xu <notifications@github.com> wrote:

> In micronota/bfillings/prodigal.py
> <https://github.com/biocore/micronota/pull/75#discussion_r61507768>:
>
> > +
> > +def _parse_single_record(chunks):
> > +    '''Parse single record of Prodigal GenBank output.
> > +
> > +    Parameters
> > +    ----------
> > +    chunks : list of str
> > +        a list of lines of the record to parse.
> > +
> > +    Yields
> > +    ------
> > +    dict passable to ``skbio.metadata.IntervalMetadata``
> > +    '''
> > +    # get the head line
> > +    head = chunks[0]
> > +    _, description = head.split(None, 1)
>
> dude, this is a common use of split. 'a b\tc'.split(None, 1) -> ['a',
> 'b\tc']
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/micronota/pull/75/files/a8ec31dbafeea33c2e04865a16f4c183f4640bdd#r61507768>
>